src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,128 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> oldNode = getNode(index);
+        Node<T> newNode = new Node<>(oldNode.prev, value, oldNode);
+        if (index == 0) {
+            head = newNode;
+        } else {
+            oldNode.prev.next = newNode;
+        }
+        oldNode.prev = newNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = getNode(index);
+        T oldValue = newNode.value;
+        newNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> oldNode = getNode(index);
+        T value = oldNode.value;
+        if (index == 0 && oldNode.next != null) {
+            head = oldNode.next;
+        } else if (oldNode.next != null && oldNode.prev != null) {
```

Comment:
what about when oldNode.next != null and oldNOde.prev == null? (that's your tail)